cli/commands/terraform/helper/aws_credentials.go

```diff
@@ -0,0 +1,106 @@
+package helper
+
+import (
+	"context"
+	"encoding/json"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/gruntwork-io/go-commons/errors"
+	"github.com/gruntwork-io/terragrunt/aws_helper"
+	"github.com/gruntwork-io/terragrunt/options"
+	"github.com/gruntwork-io/terragrunt/shell"
+)
+
+// AssumeRoleAndUpdateEnvIfNecessary assumes an IAM role by the different methods:
+// 1. attempting to run provider command if `--terragrunt-auth-provider-cmd` specified
+// 2. attempting to make API calls to Amazon STS if `--terragrunt-iam-role` specified
+// and finally setting the credentials we get back to `opts.Env`.
+func AssumeRoleAndUpdateEnvIfNecessary(ctx context.Context, opts *options.TerragruntOptions) error {
+	creds, err := getAWSCredentialsByRunningProviderCommand(ctx, opts)
+	if err != nil {
+		return err
+	}
+
+	if apiCreds, err := getAWSCredentialsByMakingAPICallsToAmazonSTS(opts); err != nil {
+		return err
+	} else if apiCreds != nil {
+		if creds != nil {
+			opts.Logger.Warnf("AWS credentials obtained using the %s command are overwritten by credentials obtained by assuming the %q role.", opts.AuthProviderCmd, opts.IAMRoleOptions.RoleARN)
+		}
+		creds = apiCreds
+	}
+
+	if creds == nil {
+		return nil
+	}
+
+	opts.Env["AWS_ACCESS_KEY_ID"] = aws.StringValue(creds.AccessKeyId)
+	opts.Env["AWS_SECRET_ACCESS_KEY"] = aws.StringValue(creds.SecretAccessKey)
+	opts.Env["AWS_SESSION_TOKEN"] = aws.StringValue(creds.SessionToken)
+	opts.Env["AWS_SECURITY_TOKEN"] = aws.StringValue(creds.SessionToken)
+
+	return nil
+}
+
+func getAWSCredentialsByRunningProviderCommand(ctx context.Context, opts *options.TerragruntOptions) (*sts.Credentials, error) {
+	command := opts.AuthProviderCmd
+	if command == "" {
+		return nil, nil
+	}
+
+	var args []string
+	if parts := strings.Fields(command); len(parts) > 1 {
+		command = parts[0]
+		args = parts[1:]
+	}
+
+	output, err := shell.RunShellCommandWithOutput(ctx, opts, "", true, false, command, args...)
+	if err != nil {
+		return nil, err
+	}
+
+	if output.Stdout == "" {
+		return nil, errors.Errorf("the command %s completed successfully, but the response does not contain json string", command)
+	}
+
+	type credentials struct {
```

Comment:
You can parse the response as a `map[string]string` and set the environment variables to what it tells you. This shouldn't have any special awareness of the fact that it might be used for AWS.